addons/web/static/src/services/router_service.js

```diff
@@ -55,105 +55,125 @@ function getRoute() {
   return { pathname, search: searchQuery, hash: hashQuery };
 }
 
-/**
- * @param {function} getCurrent function that returns the current route
- * @returns {function} function to compute the next hash
- */
-export function makePreProcessQuery(getCurrent) {
-  const lockedKeys = new Set();
-  return (hash) => {
-    const newHash = {};
-    Object.keys(hash).forEach((key) => {
-      if (lockedKeys.has(key)) {
-        return;
-      }
-      const k = key.split(" ");
-      let value;
-      if (k.length === 2) {
-        value = hash[key];
-        key = k[1];
-        if (k[0] === "lock") {
-          lockedKeys.add(key);
-        } else if (k[0] === "unlock") {
-          lockedKeys.delete(key);
-        } else {
-          return;
-        }
-      }
-      newHash[key] = value || hash[key];
-    });
-    const current = getCurrent();
-    Object.keys(current.hash).forEach((key) => {
-      if (lockedKeys.has(key) && !(key in newHash)) {
-        newHash[key] = current.hash[key];
-      }
-    });
-    return newHash;
-  };
+function applyLocking(lockedKeys, hash, currentRoute, lock=false) {
+  const newHash = {};
+  Object.keys(hash).forEach((key) => {
+    if (!lock && lockedKeys.has(key)){ // forbid implicit override of key
+      return;
+    }
+    if (lock === 'on') {
+      lockedKeys.add(key);
+    } else if (lock === 'off') {
+      lockedKeys.delete(key);
+    }
+    newHash[key] = hash[key];
+  });
+
+  Object.keys(currentRoute.hash).forEach((key) => {
+    if (lockedKeys.has(key) && !(key in newHash)) {
+      newHash[key] = currentRoute.hash[key];
+    }
+  });
+  return newHash;
 }
 
-function makeRouter(env) {
+export function routerSkeleton(env, {getRoute, historyPush, historyReplace, redirect}) {
   let bus = env.bus;
   let current = getRoute();
-  window.addEventListener("hashchange", () => {
+  const lockedKeys = new Set();
+
+  bus.on('hashchange', null, () => {
     current = getRoute();
-    bus.trigger("ROUTE_CHANGE");
+    bus.trigger('ROUTE_CHANGE');
   });
 
-  function doPush(mode = "push", route) {
-    if (!shallowEqual(route.hash, current.hash)) {
-      const url = location.origin + routeToUrl(route);
-      if (mode === "push") {
-        window.history.pushState({}, url, url);
-      } else {
-        window.history.replaceState({}, url, url);
-      }
+  function computeNewRoute(hash, replace, currentRoute) {
+    hash = sanitizeHash(hash);
+    if (!replace) {
+      hash = Object.assign({}, currentRoute.hash, hash);
     }
-    current = getRoute();
+    if (!shallowEqual(hash, currentRoute.hash)) {
+      return Object.assign({}, currentRoute, { hash });
+    }
+    return false;
   }
 
-  function getCurrent() {
-    return current;
+  function replayPushes(allCallsArgs) {
```

Comment:
replay sounds like we already considered them, which we don't. Moreover, this function only computes stuff, if has no side-effect. So I suggest something like: aggregatePushes or something like that.